### PR TITLE
added document to describe support for naming hierarchy

### PR DESCRIPTION
**Pull request by yvesgoeleven**

### CODE REVIEW

**Comment by yvesgoeleven** [comment]
@Particular/azure-service-bus-maintainers @Particular/docs-maintainers please review

File: nservicebus/azure-service-bus/namespace-hierarchy.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+https://mynamespace.servicebus.Windows.net/production/tenant1/operations
+https://mynamespace.servicebus.Windows.net/acceptance/tenant1/sales
+https://mynamespace.servicebus.Windows.net/acceptance/tenant1/operations
+```
+
+For more information about this feature refer to the [Azure Service Bus MSDN documentation on 'Addressing and Protocol'](https://msdn.microsoft.com/en-us/library/azure/hh780781.aspx)
+
+## Positioning an endpoint in the hierarchy
+
+NServiceBus provides the capability to an endpoint to register it's Azure Service Bus transport resources inside a namespace hierarchy by replacing part of the addressing logic, the composition strategy. This strategy is responsible for composing the path to an entity in the namespace. Out of the box, the transport comes with a `HierarchyComposition` strategy that allows to specify a lambda expression that can calculate the path for each entity.
```
Comment (melkio) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: Should we describe the default configured strategy and how to override default configuration?

**Comment by SimonCropp** [comment]
needs a rebase

File: nservicebus/azure-service-bus/namespace-hierarchy.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+At the core of the Azure Service Bus service, there is a Service Registry which tracks the location of each queue, topic, relay or eventhub in the service. This Service Registry provides a DNS integrated hierarchical naming system, that has a root entry point (called a namespace) at a uri with the following scheme.
+
+`https://{serviceNamespace}.servicebus.Windows.net/{path}`
+
+The {path} parameter can be specified at any depth, allowing the hierarchy to be extended with application specific sub trees. This is very usefull for large systems to group, manage and secure resources more efficiently.
+
+for example:
+
+```
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: can u append a `no-highlight`. if not hljs will guess the langauge and it is not too good at it
more info: https://github.com/Particular/docs.particular.net#inline-code

File: nservicebus/azure-service-bus/namespace-hierarchy.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+For more information about this feature refer to the [Azure Service Bus MSDN documentation on 'Addressing and Protocol'](https://msdn.microsoft.com/en-us/library/azure/hh780781.aspx)
+
+## Positioning an endpoint in the hierarchy
+
+NServiceBus provides the capability to an endpoint to register it's Azure Service Bus transport resources inside a namespace hierarchy by replacing part of the addressing logic, the composition strategy. This strategy is responsible for composing the path to an entity in the namespace. Out of the box, the transport comes with a `HierarchyComposition` strategy that allows to specify a lambda expression that can calculate the path for each entity.
+
+Snippet: asb-hierarchy-composition
+
+Note that the path generator is a lambda function, so it will be invoked each time the transport wants to determine the location of a given entity. This function should return the path to the entity, without appending the entity name itself and without a trailing slash.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: "should"? do we mean "must return"?

File: nservicebus/azure-service-bus/namespace-hierarchy.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+## Implementing a custom composition strategy
+
+If for performance or other reasons a lambda function invocation would not be ideal, it is also possible to create a custom composition strategy, by implementing `ICompositionStrategy`.
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: "for performance"? under what circumstances would the perf of the lambda have any more impact than a custom implementation?

File: nservicebus/azure-service-bus/namespace-hierarchy.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+component: ASB
+tags:
+- Cloud
+- Azure
+- Transports 
+---
+
+At the core of the Azure Service Bus service, there is a Service Registry which tracks the location of each queue, topic, relay or eventhub in the service. This Service Registry provides a DNS integrated hierarchical naming system, that has a root entry point (called a namespace) at a uri with the following scheme.
+
+`https://{serviceNamespace}.servicebus.Windows.net/{path}`
```
Comment (SimonCropp) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: single ticks are for inline code. use triple ticks with a `no-highlight`

Comment (SimonCropp) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: should "Windows" be upper case?

File: nservicebus/azure-service-bus/namespace-hierarchy.md (hunk 4 above)
Comment (yvesgoeleven) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: Anything that involves IO

File: nservicebus/azure-service-bus/namespace-hierarchy.md (hunk 1 above)
Comment (yvesgoeleven) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: I'll mention the default, how to override is shown at the end of the document already, no need to repeat

File: nservicebus/azure-service-bus/namespace-hierarchy.md (hunk 4 above)
Comment (SimonCropp) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: But in this context why would an interface handle it better than a lambda

Comment (yvesgoeleven) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: Define 'better' :p it's just easier to hide implementation details such as caching behind an interface implementation then inside a lambda

File: nservicebus/azure-service-bus/namespace-hierarchy.md (hunk 5 above)
Comment (yvesgoeleven) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: changed both

File: nservicebus/azure-service-bus/namespace-hierarchy.md (hunk 3 above)
Comment (yvesgoeleven) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: changed

File: nservicebus/azure-service-bus/namespace-hierarchy.md (hunk 2 above)
Comment (yvesgoeleven) [inline nservicebus/azure-service-bus/namespace-hierarchy.md]: done

**Comment by yvesgoeleven** [comment]
rebased

**Pull request merged**